### PR TITLE
Core files added, along with example case for Figure 1.

### PR DESCRIPTION
**Pull request by thvasilo**
A couple of TODOs left that need reviewing. Also please take a look at docstrings for omissions and inaccuracies.

Up next: More examples that are going to require pulling data sources and pre-processing with scripts (Artists) as well as bigram/pair parsing (Words).

### CODE REVIEW

File: src/main/scala/se/sics/concepts/core/CGraph.scala [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+
+  def clustersByConnectivityGraphX(): RDD[(Long, Long)] = {
+    val edges: RDD[Edge[Double]] = directedSimilarityEdges.map{case ((i, j), w) => Edge(i, j, w)}
+    val similarityGraph: Graph[String, Double] = Graph(vertexIndexStrings, edges)
+    ConnectedComponents.run(similarityGraph).vertices
+  }
+
+  // Label propagation clustering
+  // TODO(thvasilo): Remove for release?
+  def clustersBySimilarity(numberOfIterations: Int, useWeights: Boolean, saturation: Double = 1.0): RDD[(Long, Long)] = {
```
Comment (thvasilo) [inline src/main/scala/se/sics/concepts/core/CGraph.scala:252]: Remove

File: src/main/scala/se/sics/concepts/examples/FigureOne.scala [hunk 2]
```diff
@@ -0,0 +1,55 @@
+package se.sics.concepts.examples
+
+import org.apache.spark.{SparkContext, SparkConf}
+import se.sics.concepts.core.{CGraph, GraphBuilder, CKryoRegistrator}
+
+/**
+ * Running this program will create a json file as output, TODO(thvasilo): where?
+ * which you can visualize using D3.js TODO(thvasilo): Will we include vis tool?
```
Comment (thvasilo) [inline src/main/scala/se/sics/concepts/examples/FigureOne.scala:8]: Output file in root, don't include vis tool (beyond scope of this repo), but possibly link to website where vis is available.

File: src/main/scala/se/sics/concepts/core/KryoRegistrator.scala [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+import com.esotericsoftware.kryo.Kryo
+import org.apache.spark.serializer.KryoRegistrator
+
+/** Enable Kryo serialization of classes here */
+class CKryoRegistrator extends KryoRegistrator {
+  override def registerClasses(kryo: Kryo) {
+    kryo.register(classOf[CGraph])
+    kryo.register(classOf[GraphBuilder])
+  }
+}
```
Comment (thvasilo) [inline src/main/scala/se/sics/concepts/core/KryoRegistrator.scala:12]: Add newline at EOF

File: src/main/scala/se/sics/concepts/core/CGraph.scala [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+    if (inDegreeThreshold == Long.MaxValue) pruneByWeightThresholds
+    else pruneByInDegree
+
+  edgeWeightTuples.unpersist()
+
+  lazy val prunedEdges = edgesAndWeightSums._1.cache
+  lazy val discardedEdgeWeightSums = edgesAndWeightSums._2.cache
+  lazy val remainingEdgeWeightSums = edgesAndWeightSums._3.cache
+
+  // (i, j, sij, ri, rj, di, dj)
```
Comment (olofg) [inline src/main/scala/se/sics/concepts/core/CGraph.scala:144]: Remove.

File: src/main/scala/se/sics/concepts/core/CGraph.scala [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+  /**
+   * Calculates remaining and discarded weight sums for each pair after the pruning is done.
+   *
+   * Specifically the resulting RDD has the format (i, j, sij, ri, rj, di, dj) where:
+   * i and j are the vertex pair ids; sij is the contribution to the L1 norm where i and j share
+   * neighbors; ri, rj are the remaining weight sums of vertex i and j respectively and di, dj
+   * are the discarded weight sums for i and j respectively.
+   *
+   */
+  lazy val sharedTermsWithSums: RDD[(Long, Long, Double, Double, Double, Double, Double)] = {
```
Comment (olofg) [inline src/main/scala/se/sics/concepts/core/CGraph.scala:154]: Possibly put in relativeL1NormWithErrorBound.

File: src/main/scala/se/sics/concepts/core/CGraph.scala [hunk 6]
```diff
@@ -0,0 +1,409 @@
...
+    // Collect incoming edges per vertex
+    val inEdgesPerVertex = prunedEdges
+      .map { case ((i, j), w) => (j, (i, w))}
+      // TODO(thvasilo): Perhaps the number of partitions should be proportional to the number of vertices.
```
Comment (olofg) [inline src/main/scala/se/sics/concepts/core/CGraph.scala:158]: Remove comment.

File: src/main/scala/se/sics/concepts/core/CGraph.scala [hunk 7]
```diff
@@ -0,0 +1,409 @@
...
+    sharedTermsWithSums
+      .map { case (i, j, sij, ri, rj, di, dj) => {
+        val tot = ri + rj + di + dj
+        ((i, j), (sij + tot) / tot, (di + dj) / tot)
+        }
+      }
+  }
+
+  /** Similarities, where S(i, j) is one subtracted by the relative L1 norm */
+  // TODO: incorporate error bound here as well
```
Comment (olofg) [inline src/main/scala/se/sics/concepts/core/CGraph.scala:202]: Remove.

File: src/main/scala/se/sics/concepts/core/CGraph.scala [hunk 8]
```diff
@@ -0,0 +1,409 @@
...
+  lazy val simEdges: RDD[((Long, Long), Double)] = {
+    relativeL1NormWithErrorBound.map { case ((i, j), relL1, maxError) => ((i, j), 1.0 - relL1)}
+  }
+
+  def directedSimilarityEdges = simEdges.flatMap{case ((i, j), w) =>  Iterable(((i, j), w), ((j, i), w))}
+
+  // TODO: Duplicates some code here - implement separate index to label function instead
```
Comment (olofg) [inline src/main/scala/se/sics/concepts/core/CGraph.scala:209]: Remove.

File: src/main/scala/se/sics/concepts/core/CGraph.scala [hunk 9]
```diff
@@ -0,0 +1,409 @@
...
+    // Map indices to labels
+    simEdges
+      .map { case ((i, j), wij) => (i, (j, wij)) }
+      .join(vertexIndexStrings)
+      .map { case (i, ((j, wij), si)) => (j, (si, wij)) }
+      .join(vertexIndexStrings)
+      .map { case (j, ((si, wij), sj)) => ((si, sj), wij) }
+  }
+
+  // Test of community detection using GraphX' label propagation algorithm
```
Comment (olofg) [inline src/main/scala/se/sics/concepts/core/CGraph.scala:236]: Remove.

File: src/main/scala/se/sics/concepts/core/CGraph.scala [hunk 10]
```diff
@@ -0,0 +1,409 @@
...
+  // Note: Current GraphX implementation does not take edge weights into consideration
+  def clustersBySimilarityGraphX(numberOfIterations: Int): RDD[(Long, Long)] = {
+    val edges: RDD[Edge[Double]] = directedSimilarityEdges.map{case ((i, j), w) => Edge(i, j, w)}
+    val similarityGraph: Graph[String, Double] = Graph(vertexIndexStrings, edges)
+    LabelPropagation.run(similarityGraph, numberOfIterations).vertices
+  }
+
+  def clustersByConnectivityGraphX(): RDD[(Long, Long)] = {
```
Comment (olofg) [inline src/main/scala/se/sics/concepts/core/CGraph.scala:244]: Remove.

File: src/main/scala/se/sics/concepts/core/CGraph.scala (hunk 1 above)
Comment (olofg) [inline src/main/scala/se/sics/concepts/core/CGraph.scala]: Remove.

File: src/main/scala/se/sics/concepts/core/CGraph.scala [hunk 11]
```diff
@@ -0,0 +1,409 @@
...
+    val file = new PrintWriter(new OutputStreamWriter(new BufferedOutputStream(new FileOutputStream(fileName)), "UTF-8"), false)
+    file.write(Json.prettyPrint(jsonGraph))
+    file.close()
+  }
+
+  /**
+   * Print cluster assignments to stdout
+   * @param clusterAssignments The cluster assignments RDD as returned by [[CGraph.clustersByConnectivityGraphX()]]
+   */
+  def printClusters(clusterAssignments: RDD[(Long, Long)]): Unit = {
```
Comment (olofg) [inline src/main/scala/se/sics/concepts/core/CGraph.scala:352]: Remove.

File: src/main/scala/se/sics/concepts/core/GraphBuilder.scala [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+    bigramsRDD.saveAsTextFile(graphFilePrefix + "-bigrams")
+    val (vertices, edges) = vtxEdgesFromBigram(bigramsRDD)
+
+    // Write edges and vertices to text file
+    edges.saveAsTextFile(graphFilePrefix + "-edges")
+    vertices.saveAsTextFile(graphFilePrefix + "-vertices")
+
+    println(s"#vertices:    ${vertices.count}")
+    println(s"#edges:       ${edges.count}")
+    // println(s"#occurrences: ${totalVertexCount}")
```
Comment (olofg) [inline src/main/scala/se/sics/concepts/core/GraphBuilder.scala:84]: Remove.

File: src/main/scala/se/sics/concepts/core/CGraph.scala (hunk 10 above)
Comment (thvasilo) [inline src/main/scala/se/sics/concepts/core/CGraph.scala:244]: Removing this would mean we do no clustering for FigureOne example. Is this what we want?

**Pull request merged**